core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -335,9 +339,10 @@ private[spark] class TaskSchedulerImpl(
     for (i <- 0 until shuffledOffers.size) {
       val execId = shuffledOffers(i).executorId
       val host = shuffledOffers(i).host
-      if (availableCpus(i) >= CPUS_PER_TASK) {
+      if (availableCpus(i) >= CPUS_PER_TASK &&
+        resourceMeetTaskRequirements(availableResources(i))) {
         try {
-          for (task <- taskSet.resourceOffer(execId, host, maxLocality)) {
+          for (task <- taskSet.resourceOffer(execId, host, maxLocality, availableResources(i))) {
             tasks(i) += task
             val tid = task.taskId
             taskIdToTaskSetManager.put(tid, taskSet)
```

Comment:
I think its a bit out of place that we are decrementing the addresses  available in the resourceOffer function since everything else does its bookkeeping right here.  I don't think that is a blocker for this though and the way

Reply:
I just feel if we only bookkeep by count here and actually allocate the addresses in SchedulerBackend then we need to pay slightly more attention to ensure these two places are synchronized. After all I'm neutral here and would be happy to go another way if someone feels strongly about this.

Reply:
I like @tgravescs 's proposal or at least putting an inline comment after L352 saying where we update the addresses.

Reply:
I'd be ok with a comment for now.  I wasn't suggesting only book keeping by count, I meant do the actual acquire addresses here, but you would have to make other changes for that to happen.  
One thing we could do is make a separate function in TaskSchedulerImpl for the resource (including cpu's) accounting and then have TaskSetManager.resourceOffer call into that to acquire the resources.
Another option is to have the TaskSetManager get some addresses but only do the actual accounting to remove them from the available here.